doc/developer-guide/internal-libraries/RPC.en.rst

```diff
@@ -0,0 +1,403 @@
+
+When data is actually sent over a connection it must first be seralized. This is the general serialization mechanism for RPC communcation. 
+
+Generally, for remote clients sending messages to |TServer|, the message is serialized using remote RPC APIs. The serialized message is sent to |TManager| and |TManager| then simply relays the message to |TServer|. |TServer| eventually unserializes the message. 
+
+Details for how |TManager| and |TServer| communicate are documented in the next section.
+
+Marshalling:
+============
+
+.. function:: ssize_t MgmtMarshall::mgmt_message_marshall(void *buf, size_t remain, const MgmtMarshallType *fields, unsigned count, ...)
+
+   Variable argument wrapper for ``mgmt_message_marshall_v``. Allows for different datatypes to be marshalled together as long as a field is specified for each data object. Arguments should be references to objects to be marshalled. 
+
+.. function:: ssize_t MgmtMarshall::mgmt_message_marshall_v(void *buf, size_t remain, const MgmtMarshallType *fields, unsigned count, va_list ap)
+
+   This function goes through all the data objects and serializes them together into a buffer. Based on the field, the number of bytes is determined and if there is enough space, it is written into the buffer. 
+
+Unmarshalling:
+==============
+
+.. function:: ssize_t MgmtMarshall::mgmt_message_parse(const void *buf, size_t len, const MgmtMarshallType *fields, unsigned count, ...)
+
+   Variable argument wrapper for ``mgmt_message_parse_v``. Reference to data object to store unmarshalled message needed for variable arguements. 
+
+.. function:: ssize_t MgmtMarshall::mgmt_message_parse_v(const void *buf, size_t len, const MgmtMarshallType *fields, unsigned count, va_list ap)
+   
+   This function parses all the serialized. Based on the field, the number of bytes to be read is determined and copied into a ``MgmtMarshallAnyPtr``.
+
+===================
+Local Serialization
+===================
+
+State Machine diagrams for |TServer| and |TManager| RPC communication. Messages between |TManager| and |TServer| begin with a ``MgmtMessageHdr`` followed by the data. 
+
+.. image:: ../../uml/images/RPC-states.svg
+   :align: center
+
+.. class:: MgmtMessageHdr
+
+   .. member:: int msg_id 
+
+      ID for the event or signal to be sent.
+
+   .. member:: int data_len
+
+      Length in bytes of the message. 
+
+A RPC message is sent as a ``MgmtMessageHdr`` followed by the serialized data in bytes. 
+
+Read:
+=====
+
+1. Read the MgmtMessageHdr:
+
+   a. ``MgmtUtils.cc::mgmt_read_pipe(int fd, char *buf, int bytes_to_read)`` calls,
+
+   #. ``ink_sock.cc::read_socket(int s, char *buffer, int length)`` casts char* buffer to void*, calls, 
+
+   #. ``read(int fd, void *buf, size_t count)`` <- a C system call.
+
+#. Read the content:
+
+   a. data length is given in the header, 
+
+   #. read data_len bytes with the same proceedure as the MgmtMessageHdr.
+
```

Comment:
Instead of this list, you might try something like
```
.. uml::

   partition "Read MgmtMessageHdr" {
   :mgmt_read_pipe();
   -> calls;
   :read_socket();
   -> calls;
   :read();
   note right: System call
   }

   partition "Read message content" {
   :Get content size from header;
   :mgmt_read_pipe();
   -> calls;
   :read_socket();
   -> calls;
   :read();
   note right: System call
   }
```
This is "Activity Diagram" (beta) in PlantUML.